modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/PlannerHelper.java

```diff
@@ -73,8 +75,9 @@ public static IgniteRel optimize(SqlNode sqlNode, IgnitePlanner planner, IgniteL
             // Convert to Relational operators graph.
             RelRoot root = planner.rel(sqlNode);
 
-            planner.setDisabledRules(HintUtils.options(root.rel, extractRootHints(root.rel),
-                HintDefinition.DISABLE_RULE));
+            root = addExternalHint(root, planner);
+
+            planner.setDisabledRules(HintUtils.options(root.rel, extractRootHints(root.rel), HintDefinition.DISABLE_RULE));
```

Comment:
We use two different mechanisms to disable rules (planner.setDisabledRules and HintStrategy.excludedRules) can we use only one of them?

Reply:
`planner.setDisabledRules()` disables rules for the whole query, including subqueries and unions. `HintStrategy's excludedRules` is able to just skip rule for certain RelNode (in `RelOptRuleCall#isRuleExcluded()`). If we process this hint as a disabled rules, tests like `testDisabledCommutingOfJoinInputsInSubquery()` or `testUnions()` won't work correctly. Join order will be disabled for all the joins.